pkg/network/kube-proxy.go

```diff
@@ -0,0 +1,175 @@
+	}
+
+	if conf.KubeProxyConfig.ProxyArguments == nil {
+		conf.KubeProxyConfig.ProxyArguments = map[string][]string{}
+	}
+
+	if conf.KubeProxyConfig.ProxyArguments["metrics-bind-address"] == nil {
+		conf.KubeProxyConfig.ProxyArguments["metrics-bind-address"] = []string{"0.0.0.0"}
+	}
+
+	if conf.KubeProxyConfig.ProxyArguments["metrics-port"] == nil {
+		conf.KubeProxyConfig.ProxyArguments["metrics-port"] = []string{"9101"}
+	}
+
+	if len(conf.ClusterNetwork) == 1 && conf.KubeProxyConfig.ProxyArguments["cluster-cidr"] == nil {
+		conf.KubeProxyConfig.ProxyArguments["cluster-cidr"] = []string{conf.ClusterNetwork[0].CIDR}
+	}
+}
+
+func RenderKubeProxy(conf *operv1.NetworkSpec, manifestDir string) ([]*uns.Unstructured, error) {
+	kpc := conf.KubeProxyConfig
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["ReleaseVersion"] = os.Getenv("RELEASE_VERSION")
+	data.Data["KubeProxyImage"] = os.Getenv("KUBE_PROXY_IMAGE")
+	data.Data["KUBERNETES_SERVICE_HOST"] = os.Getenv("KUBERNETES_SERVICE_HOST")
+	data.Data["KUBERNETES_SERVICE_PORT"] = os.Getenv("KUBERNETES_SERVICE_PORT")
+
+	manifests, err := render.RenderDir(filepath.Join(manifestDir, "kube-proxy"), &data)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to render manifests")
+	}
+
+	// Insert kube proxy args
+	args := []string{
+		"--bind-address", kpc.BindAddress,
+	}
+
+	// Sort args alphabetically, just so we can test
+	argNames := []string{}
+	for argName := range kpc.ProxyArguments {
+		argNames = append(argNames, argName)
+	}
+	sort.Strings(argNames)
+
+	for _, argName := range argNames {
+		for _, val := range kpc.ProxyArguments[argName] {
+			args = append(args, "--"+argName, val)
+		}
+	}
+
+	found := false
+	for _, manifest := range manifests {
+		if manifest.GetAPIVersion() == "apps/v1" && manifest.GetKind() == "DaemonSet" && manifest.GetName() == "openshift-kube-proxy" {
+			found = true
+			containers, ok, err := uns.NestedSlice(manifest.Object, "spec", "template", "spec", "containers")
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot inject kube-proxy args")
+			}
+			if !ok || len(containers) != 1 {
+				return nil, errors.New("cannot inject kube-proxy args: no containers")
+			}
+			container, ok := (containers[0]).(map[string]interface{})
+			if !ok {
+				return nil, errors.New("cannot inject kube-proxy args: bad cast")
+			}
+
+			existingArgs, ok, err := uns.NestedStringSlice(container, "command")
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot inject kube-proxy args")
+			}
+			if !ok {
+				return nil, errors.New("cannot inject kube-proxy args: no command")
+			}
+
+			existingArgs = append(existingArgs, args...)
+			err = uns.SetNestedStringSlice(container, existingArgs, "command")
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot inject kube-proxy args")
+			}
+
+			containers[0] = container
+
+			err = uns.SetNestedSlice(manifest.Object, containers, "spec", "template", "spec", "containers")
```

Comment:
it wouldn't be easier to serialize the args to proper yaml and then template them in?

Reply:
It's tough to control exactly how yaml will serialize stuff. I wasn't comfortable it so that the templated yaml could always safely be inserted.

This was ugly to figure out but is otherwise correct.